README.md
[![wercker status](https://app.wercker.com/status/380ff799a37ab2e1bb09a8ae5200ca99/s/master "wercker status")](https://app.wercker.com/project/byKey/380ff799a37ab2e1bb09a8ae5200ca99) [![Coverage Status](https://coveralls.io/repos/github/rafaelverger/create-git-changelog/badge.svg?branch=issue9)](https://coveralls.io/github/rafaelverger/create-git-changelog?branch=issue9)

# generate-git-changelog
NPM module to run generate git changelog based on package.json version changes


CHANGELOG.md
# CHANGELOG

## 0.2.0
- [FIX #21] Handle semver beta versions.
- [FIX #20] Fix sort versions method.
- [FIX #19] Fix indentation on changelog.
- [FIX #13] Add arg to cli to use only tagged commits.
- Avoid split rule splitting its own commit patches
  Related #16
- Forcing tag info to show on commits list
- Allow useTag parameter to only get version from tagged commits.
  Related to #13
- Update parser to get commit's tags
- [FIX #17] Improve regex to identify issue numbers.
- [FIX #15] Fix changelogs with only unpublished changes.
- [FIX #16] Fix extraction of commit's first line.
- [FIX #14] Remove unnecessary files from pkg.
- [FIX #12] Create bin file.

## 0.1.0
- [FIX #9] Integrate with coveralls.
- Add tests to create script
- Create gitignore
- Ignoring messages on commits that only updates pkg version
- Add more tests to parseCommit with resolved issues
- Impoving generator script
- [FIX #7] Using new info to filter significant commits.
- Enhancing change infos while parsing commits. rel #7
- Add wercker badge to README
- [FIX #3] Using wercker as CI.
- [FIX #4] Build changes using commits msg and grouped by version.
- [FIX #1] Create method to parse commit text.

## 0.0.0
- Add package.json
- Initial commit


util.test.js
const util = require('./util');

test('parseCommit', () => {
  const commitText = `
commit 0808ee7ab79b3d6b8740375b14e4567e827d3ccf (tag: v0.0.1, HEAD -> issue1, origin/master, master)
Author: Anon <anon@noone.knows>
Date:   2018-10-31 18:35:27 -0300

    Initial commit

    Double line

diff --git a/README.md b/README.md
new file mode 100644
index 0000000..0361a2c
--- /dev/null
+++ b/README.md
@@ -0,0 +1,2 @@
+# generate-git-changelog
+NPM module to run generate git changelog based on package.json version changes
`.trim();
  const obj = util.parseCommit(commitText);
  expect(obj).toMatchObject({
    id: '0808ee7ab79b3d6b8740375b14e4567e827d3ccf',
    tag: 'v0.0.1',
    author: 'Anon <anon@noone.knows>',
    date: new Date('2018-10-31 18:35:27 -0300'),
    message: 'Initial commit\nDouble line',
    patch: {
      'README.md': [
        'diff --git a/README.md b/README.md',
        'new file mode 100644',
        'index 0000000..0361a2c',
        '--- /dev/null',
        '+++ b/README.md',
        '@@ -0,0 +1,2 @@',
        '+# generate-git-changelog',
        '+NPM module to run generate git changelog based on package.json version changes',
      ]
    },
    ignoreMessage: false,
  })
});

test('parseCommit multiple diffs', () => {
  const commitText = `
commit 0808ee7ab79b3d6b8740375b14e4567e827d3ccf
Author: Anon <anon@noone.knows>
Date:   2018-10-31 18:35:27 -0300

    Initial commit

    Double line

diff --git a/README.md b/README.md
new file mode 100644
index 0000000..0361a2c
--- /dev/null
+++ b/README.md
@@ -0,0 +1,2 @@
+# generate-git-changelog
+NPM module to run generate git changelog based on package.json version changes

diff --git a/package.json b/package.json
new file mode 100644
index 0000000..b9a6faa
--- /dev/null
+++ b/package.json
@@ -0,0 +1,28 @@
+{
+  "name": "create-git-changelog",
+  "version": "0.0.0",
+  "description": "Automate your changelog update based on repo commits",
+  "main": "index.js",
+  "scripts": {
+    "test": "test"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/rafaelverger/create-git-changelog.git"
+  },
+  "keywords": [
+    "changelog",
+    "git",
+    "changes",
+    "update",
+    "automate",
+    "generate"
+  ],
+  "author": "rafaelverger",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/rafaelverger/create-git-changelog/issues"
+  },
+  "homepage": "https://github.com/rafaelverger/create-git-changelog#readme",
+  "engines": { "node" : ">=4" }
+}
`.trim();
  const obj = util.parseCommit(commitText);
  expect(obj).toMatchObject({
    id: '0808ee7ab79b3d6b8740375b14e4567e827d3ccf',
    tag: null,
    author: 'Anon <anon@noone.knows>',
    date: new Date('2018-10-31 18:35:27 -0300'),
    message: 'Initial commit\nDouble line',
    patch: {
      'README.md': [
        'diff --git a/README.md b/README.md',
        'new file mode 100644',
        'index 0000000..0361a2c',
        '--- /dev/null',
        '+++ b/README.md',
        '@@ -0,0 +1,2 @@',
        '+# generate-git-changelog',
        '+NPM module to run generate git changelog based on package.json version changes',
      ],
      'package.json': [
        'diff --git a/package.json b/package.json',
        'new file mode 100644',
        'index 0000000..b9a6faa',
        '--- /dev/null',
        '+++ b/package.json',
        '@@ -0,0 +1,28 @@',
        '+{',
        '+  "name": "create-git-changelog",',
        '+  "version": "0.0.0",',
        '+  "description": "Automate your changelog update based on repo commits",',
        '+  "main": "index.js",',
        '+  "scripts": {',
        '+    "test": "test"',
        '+  },',
        '+  "repository": {',
        '+    "type": "git",',
        '+    "url": "git+https://github.com/rafaelverger/create-git-changelog.git"',
        '+  },',
        '+  "keywords": [',
        '+    "changelog",',
        '+    "git",',
        '+    "changes",',
        '+    "update",',
        '+    "automate",',
        '+    "generate"',
        '+  ],',
        '+  "author": "rafaelverger",',
        '+  "license": "MIT",',
        '+  "bugs": {',
        '+    "url": "https://github.com/rafaelverger/create-git-changelog/issues"',
        '+  },',
        '+  "homepage": "https://github.com/rafaelverger/create-git-changelog#readme",',
        '+  "engines": { "node" : ">=4" }',
        '+}',
      ]
    },
    ignoreMessage: false,
  })
});

test('parseCommit merged commit', () => {
  const commitText = `
commit 2d028acf8157a6626b9dd6a080810dd25e3cff3a
Merge: 53fa825 21a6349
Author: Anon <anon@noone.knows>
Date:   2018-11-01 00:05:15 -0300

    Merge pull request #6 from rafaelverger/issue3

    Using wercker as Continuous Integration service
`.trim();
  const obj = util.parseCommit(commitText);
  expect(obj).toMatchObject({
    id: '2d028acf8157a6626b9dd6a080810dd25e3cff3a',
    tag: null,
    merge: ['53fa825', '21a6349'],
    author: 'Anon <anon@noone.knows>',
    date: new Date('2018-11-01 00:05:15 -0300'),
    message: 'Merge pull request #6 from rafaelverger/issue3\nUsing wercker as Continuous Integration service',
    patch: {},
    ignoreMessage: false,
  })
});

test('parseCommit pkg version change commit', () => {
  const commitText = `
commit cc70261c7cbe458221caec5bfab14745d73cff79 (tag: v1.3.1)
Author: Anon <anon@noone.knows>
Date:   2017-04-10 17:50:05 -0700

    Version Bump v13.1: #16 fix issue where concat was used incorrectly

diff --git a/package.json b/package.json
index b952b44..602c994 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "smtpapi",
-  "version": "1.3.0",
+  "version": "1.3.1-dev.2",
   "main": "lib/main.js",
`.trim();
  const obj = util.parseCommit(commitText);
  expect(obj).toMatchObject({
    id: 'cc70261c7cbe458221caec5bfab14745d73cff79',
    author: 'Anon <anon@noone.knows>',
    date: new Date('2017-04-10 17:50:05 -0700'),
    message: 'Version Bump v13.1: #16 fix issue where concat was used incorrectly',
    patch: {
      'package.json': [
        'diff --git a/package.json b/package.json',
        'index b952b44..602c994 100644',
        '--- a/package.json',
        '+++ b/package.json',
        '@@ -1,6 +1,6 @@',
        ' {',
        '   "name": "smtpapi",',
        '-  "version": "1.3.0",',
        '+  "version": "1.3.1-dev.2",',
        '   "main": "lib/main.js",',
      ]
    },
    version: '1.3.1-dev.2',
    ignoreMessage: true,
  })
});

test('parseCommit commit with resolved issue', () => {
  const commitTexts = [
  `
commit 21a63493df3557cf4efd31fe9d902cbd33734743
Author: Anon <anon@noone.knows>
Date:   2018-11-01 00:03:01 -0300

    Using wercker as CI.

    Testing

    close #3

diff --git a/wercker.yml b/wercker.yml
new file mode 100644
index 0000000..f6c9cb5
--- /dev/null
+++ b/wercker.yml
@@ -0,0 +1,5 @@
+box: node:4
+build:
+  steps:
+    - npm-install
+    - npm-test
  `.trim(),
  `
commit 21a63493df3557cf4efd31fe9d902cbd33734743
Author: Anon <anon@noone.knows>
Date:   2018-11-01 00:03:01 -0300

    Using wercker as CI.

    Testing

    closes #3

diff --git a/wercker.yml b/wercker.yml
new file mode 100644
index 0000000..f6c9cb5
--- /dev/null
+++ b/wercker.yml
@@ -0,0 +1,5 @@
+box: node:4
+build:
+  steps:
+    - npm-install
+    - npm-test
  `.trim(),
  `
commit 21a63493df3557cf4efd31fe9d902cbd33734743
Author: Anon <anon@noone.knows>
Date:   2018-11-01 00:03:01 -0300

    Using wercker as CI.

    Testing

    fixes #3

diff --git a/wercker.yml b/wercker.yml
new file mode 100644
index 0000000..f6c9cb5
--- /dev/null
+++ b/wercker.yml
@@ -0,0 +1,5 @@
+box: node:4
+build:
+  steps:
+    - npm-install
+    - npm-test
  `.trim(),
  `
commit 21a63493df3557cf4efd31fe9d902cbd33734743
Author: Anon <anon@noone.knows>
Date:   2018-11-01 00:03:01 -0300

    Using wercker as CI.

    Testing

    fix #3

diff --git a/wercker.yml b/wercker.yml
new file mode 100644
index 0000000..f6c9cb5
--- /dev/null
+++ b/wercker.yml
@@ -0,0 +1,5 @@
+box: node:4
+build:
+  steps:
+    - npm-install
+    - npm-test
  `.trim(),
  ];
  const expected = {
    id: '21a63493df3557cf4efd31fe9d902cbd33734743',
    author: 'Anon <anon@noone.knows>',
    date: new Date('2018-11-01 00:03:01 -0300'),
    message: 'Using wercker as CI.\nTesting',
    patch: {
      'wercker.yml': [
        'diff --git a/wercker.yml b/wercker.yml',
        'new file mode 100644',
        'index 0000000..f6c9cb5',
        '--- /dev/null',
        '+++ b/wercker.yml',
        '@@ -0,0 +1,5 @@',
        '+box: node:4',
        '+build:',
        '+  steps:',
        '+    - npm-install',
        '+    - npm-test',
      ]
    },
    resolvedIssue: '3',
    ignoreMessage: false,
  }
  commitTexts.forEach(commitText => {
    const obj = util.parseCommit(commitText);
    expect(obj).toMatchObject(expected);
  });
});


index.test.js
jest.mock('child_process');

const EventEmitter = require('events');
const fs = require('fs')

const spawnMock = require('child_process').spawn;

afterAll((done) => {
  const spawn = require.requireActual('child_process').spawn;
  const cmd = spawn('git', ['checkout', '--', 'CHANGELOG.md']);
  cmd.stderr.on('data', (data) => {
    console.error(data.toString());
    done(new Error('error clearing CHANGELOG.md file'))
  });
  cmd.on('close', (code) => done());
}, 500);

function assertChangelog(commitHistory, options, done) {
  const emitter = new EventEmitter();
  emitter.stderr = new EventEmitter();
  emitter.stdout = new EventEmitter();

  spawnMock.mockReturnValueOnce(emitter);

  const watchListener = fs.watchFile('CHANGELOG.md', (curr, prev) => {
    const changelog = fs.readFileSync('CHANGELOG.md').toString();
    expect(changelog).toMatchSnapshot();
    fs.unwatchFile('CHANGELOG.md', watchListener);
    done();
  });

  require('./index')(options);

  emitter.stdout.emit('data', commitHistory);
  emitter.emit('close', 0);
}

test('changelog creation', (done) => {
  const commitHistory = fs.readFileSync('./__mock__/commit.history');
  assertChangelog(commitHistory, {}, done);
}, 6000);

test('changelog creation using more than 2 lines', (done) => {
  const commitHistory = fs.readFileSync('./__mock__/commit.history');
  assertChangelog(commitHistory, { descLines: 3 }, done);
}, 6000);

test('changelog creation with only unpublished changes', (done) => {
  const commitHistory = fs.readFileSync('./__mock__/commit_unpublished.history');
  assertChangelog(commitHistory, {}, done);
}, 6000);

test('changelog creation with tagged commits', (done) => {
  const commitHistory = fs.readFileSync('./__mock__/commit_tagged.history');
  assertChangelog(commitHistory, { useTags: true }, done);
}, 6000);


util.js

const COMMIT_HASH_REGEX = /^commit\s+([a-z0-9]{40}).*/i;
const COMMIT_TAG_REGEX = /^.*\(tag:\s?(.*?)(,.*)*\)/i;
const FIX_ISSUE_REGEX = /((fix(es)?)|(closes?))\s?#(\d+)\b/i;
const COMMIT_PROPERTIES_HANDLER = (v) => {
  switch(v.toLowerCase()){
    case 'date':
      return x => new Date(x);
    case 'merge':
      return x => x.split(' ');
    default:
      return x => x
  }
};

function getCommitHash(commitFirstLine) {
  return COMMIT_HASH_REGEX.exec(commitFirstLine)[1];
}

function findCommitTag(commitFirstLine) {
  const rs = COMMIT_TAG_REGEX.exec(commitFirstLine);
  return rs ? rs[1] : null;
}

function parseCommit(commitText) {
  const changeLines = commitText.split('\n').filter(line => line.trim());
  const messageStartLine = changeLines.findIndex(line => line.match(/^date:\s*/i)) + 1;
  const patchingDiffStartLine = Math.max(0, changeLines.findIndex(
    line => line.indexOf('diff --git') === 0
  )) || changeLines.length;

  const change = Object.assign(
    {
      // first line of commit message is its id and tag
      id: getCommitHash(changeLines[0]),
      tag: findCommitTag(changeLines[0])
    },
    changeLines.slice(1, messageStartLine).reduce((obj, line) => {
      const kv = line.split(':');
      const key = kv[0].toLowerCase().trim();
      const value = kv.slice(1).join(':').trim();
      const handler = COMMIT_PROPERTIES_HANDLER(key);
      obj[key] = handler(value);
      return obj;
    }, {}),
    {
      // after commit date line we have commit's message untill patching diff appears
      message: changeLines.slice(messageStartLine, patchingDiffStartLine).map(l => l.trim()).join('\n'),
      // if commit have patching diff it will be stored below
      patch: changeLines.slice(patchingDiffStartLine).reduce((obj, line, idx, arr) => {
        if (line.indexOf('diff --git') === 0) {
          const fname = line.replace(/^diff --git a\/.* b\//, '');
          obj[fname] = [];
          obj.current = obj[fname];
        }
        obj.current.push(line);
        if (idx === arr.length-1) {
          obj.current = null;
          delete obj.current;
        }
        return obj;
      }, {}),
    }
  );
  const packageChange = change.patch['package.json'];
  if (packageChange) {
    const versionChange = packageChange.find(diffLine => diffLine.indexOf('+  "version":') === 0);
    if (versionChange) {
      change.version = versionChange.replace(/.*?"(\d+\.\d+\.\d+(-\w+(\.\d+)?)?)".*/, '$1');
    }
  }

  const issueRelated = FIX_ISSUE_REGEX.exec(change.message)
  if (issueRelated) {
    change.resolvedIssue = issueRelated[5];
    change.message = change.message
      .replace(issueRelated[0], '')
      .replace(/\s+?(\s)/g, '$1')
      .trim();
  }

  change.ignoreMessage = !!(
    change.version &&
    Object.keys(change.patch).every(file => file.match(/(package.json)|(changelog.md)/i)) &&
    change.patch['package.json'].filter(l => l.match(/^\+[^\+]/)).length === 1
  );

  return change;
}

module.exports = {
  parseCommit
}


logger.js
const readline = require('readline');

module.exports = {
  info: function(msg) {
    this.__progress_clear();
    process.stdout.write(msg + '\n');
  },
  error: function(msg) {
    this.__progress_clear();
    process.stdout.write('[ERROR]: ' + msg + '\n');
  },
  __progress_spinner: ['|', '/', '-', '*'],
  __progress_clear: function() {
    if (this.__progress) {
      process.stdout.write('\n');
      delete this.__progress;
    }
  },
  progress: function(msg) {
    if (this.__progress !== undefined) {
      readline.clearLine(process.stdout, 0);
      readline.cursorTo(process.stdout, 0, null);
    } else {
      this.__progress = (this.__progress || 0);
    }
    const spinIdx = this.__progress % this.__progress_spinner.length;
    const spinChar = this.__progress_spinner[spinIdx];
    process.stdout.write('[' + spinChar + '] ' + msg);
    this.__progress++;
  }
};


index.js
const spawn = require('child_process').spawn;
const fs = require('fs');

const logger = require('./logger');
const parseCommit = require('./util').parseCommit;

const pkg = JSON.parse(fs.readFileSync('package.json').toString());
const currVersion = pkg.version;

const defaultOptions = { useTags: false, descLines: 2 };
module.exports = function(options) {
  logger.info('Creating changelog. Current version: ' + currVersion);
  const opts = Object.assign({}, defaultOptions, options);
  logger.info('** Commits description will be truncated in ' + opts.descLines + ' lines');
  const useTags = opts.useTags;
  if (useTags) {
    logger.info('** Using only tagged commits as version changes')
  }

  logger.info('Fetching git-log...');
  const gitHistory = spawn('git', ['log', '-p', '--date=iso', '--decorate']);
  gitHistory.stderr.on('data', (data) => {
    logger.error(data.toString());
  });

  // `\n` is necessary due to `split` in line#16
  var commitsHistory = '\n';

  gitHistory.stdout.on('data', (buffer) => {
    commitsHistory += buffer.toString();
    const logSize = parseFloat(commitsHistory.length/1024.0).toFixed(2);
    logger.progress('Fetched ' + logSize + 'Kb from git-log...');
  });


  gitHistory.on('close', (code) => {
    logger.info('git-log fetched')
    // doing (':|:'+':|:') is necessary to avoid spliting this line after commit
    const changes = commitsHistory
      .replace(/^(commit\s+([a-zA-Z0-9]{40})(\s+\([^\)]+\))?)$/gm, ':|:'+':|:$1')
      .split(':|:'+':|:')
      .slice(1);
    const totalChanges = changes.length;
    logger.info(totalChanges + ' commits were found');

    logger.info('Parsing commits...')
    const parsedChanges = changes.map((c, i) => {
      logger.progress('Parsing commits: ' + parseFloat(100.0*i/totalChanges).toFixed(2) + '%');
      return parseCommit(c);
    }).filter(change => (
      change.resolvedIssue || Object.keys(change.patch).length
    ));
    logger.info('Valid commits: ' + parsedChanges.length);

    logger.info('Sorting commits by version...');
    const unpublishedChanges = [];
    const changesTracker = parsedChanges.reduce((tracker, change, idx, arr) => {
      if ((useTags && change.tag) || (!useTags && change.version)) {
        const version = useTags ? change.tag : change.version;
        if (Object.keys(tracker).length === 1) {
          // if it's the first version found, set curr commits as unpublished version
          [].push.apply(unpublishedChanges, tracker.curr);
        }
        tracker[version] = [];
        tracker.curr = tracker[version];
      }
      tracker.curr.push(change);

      if (Object.keys(tracker).length === 1 && idx === arr.length-1) {
        // if it's the last item and no version is found, set curr commits as unpublished version
        [].push.apply(unpublishedChanges, tracker.curr);
      }
      return tracker;
    }, { curr: [] });
    changesTracker.curr = null;
    delete changesTracker.curr;

    logger.info('Writing changelog...');
    const writeVersionChanges = (version, changes) => {
      if (changes.length === 0) return [];
      return ['\n## ' + version].concat(
        changes.map(change => {
          if (change.ignoreMessage) return '';
          const resolvedIssue = change.resolvedIssue;
          return (
            '- ' +
            (resolvedIssue ? '[FIX #' + resolvedIssue + '] ' : '') +
            change.message.split(/\n/).slice(0, opts.descLines).join('\n  ')
          );
        }).filter(msg => msg)
      )
    };
    const changelog_md = ['# CHANGELOG'].concat(
      writeVersionChanges('unpublished', unpublishedChanges),
      Object.keys(changesTracker)
        .sort()
        .reverse()
        .reduce((acc, version) => acc.concat(
          writeVersionChanges(version, changesTracker[version])
        ), [])
    );

    fs.writeFileSync('CHANGELOG.md', changelog_md.join('\n') + '\n', 'utf8');
    logger.info('Changes dump into CHANGELOG.md');
  });
};
